python/tvm/relay/testing/py_converter.py

```diff
@@ -0,0 +1,616 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Utility for converting Relay code into a Python script with equivalent semantics"""
+import ast
+from ast import alias, Assign, Load, Name, NameConstant, Num, Return, Store, Str
+import re
+
+import tvm
+from tvm import relay
+from tvm.relay.adt import Constructor, Pattern
+from tvm.relay.backend import compile_engine
+from tvm.relay.expr import Expr, Function, GlobalVar, Var
+from tvm.relay.expr_functor import ExprFunctor
+
+OUTPUT_VAR_NAME = '_py_out'
+MODULE_NAME = '_mod'
+
+# corresponds to:
+#     import numpy
+#     import tvm
+#     from tvm import relay
+#     from tvm.relay.backend.interpreter import RefValue, TupleValue, TensorValue, ConstructorValue
+PROLOGUE = [
+    ast.Import([alias('numpy', None)]),
+    ast.Import([alias('tvm', None)]),
+    ast.ImportFrom('tvm', [alias('relay', None)], 0),
+    ast.ImportFrom('tvm.relay.backend.interpreter',
+                   [alias('RefValue', None),
+                    alias('TupleValue', None),
+                    alias('TensorValue', None),
+                    alias('ConstructorValue', None)],
+                   0)
+]
+
+class PythonConverter(ExprFunctor):
+    '''Functor for translating Relay programs into Python ASTs.'''
+
+    def __init__(self, mod, target) -> None:
+        super().__init__()
+        self.mod = mod
+        self.tgt = target
+        self.engine = compile_engine.get()
+        self.fun_no = 0
+        self.var_no = 0
+        self.var_map = {}
+
+
+    def convert(self, prog: Expr):
+        '''This method converts the passed Relay expression into a Python
+        AST object with equivalent semantics.
+
+        The Python AST can be executed using exec(); it can be turned
+        into text and inspected using astor.
+        '''
+        optimized = self.optimize(prog)
+
+        # start with conversion prelude (imports) and convert global defs
+        body = []
+        body += PROLOGUE
+        body += self.convert_module()
+
+        prog_body, extra_defs = self.visit(optimized)
+        body += extra_defs
+
+        # we finally must assign the final expression to the output var
+        # so it can be read after running EXEC
+        body.append(Assign([Name(OUTPUT_VAR_NAME, Store())], prog_body))
+
+        return ast.fix_missing_locations(ast.Module(body=body))
+
+
+    def optimize(self, prog: Expr):
```

Comment:
Should we add a todo here to move these passes to the pass manager?

Reply:
Ah, I haven't been in the loop on that. If the pass manager is the right way to do these, then we should use it

Reply:
It's not merged yet... The TODO just helps us to track it.